tpm2/marshalling.go

```diff
@@ -0,0 +1,237 @@
+package tpm2
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"reflect"
+)
+
+// maybe represents some type that came out of an unmarshalling process.
+// We may have the value, or we may have an error.
+type maybe[T any] struct {
+	err   error
+	value *T
+}
+
+// OK returns true if the maybe contains contents instead of an error.
+func (m maybe[_]) OK() bool {
+	return m.err == nil
+}
+
+// Unwrap unwraps the result of an unmarshalling operation.
+// Panics if the data was not unmarshalled.
+func (m maybe[T]) Unwrap() *T {
+	if m.err != nil {
+		panic(fmt.Sprintf("could not unwrap: %v", m.err))
+	}
+	return m.value
+}
+
+// CheckUnwrap unwraps the result of an unmarshalling operation.
+func (m maybe[T]) CheckUnwrap() (*T, error) {
+	if m.err != nil {
+		return nil, m.err
+	}
+	return m.value, nil
+}
+
+// asMaybe returns a maybe enclosing the given contents.
+func asMaybe[T any](t *T) maybe[T] {
+	return maybe[T]{
+		value: t,
+	}
+}
+
+// maybeNot returns a maybe with the given error.
+func maybeNot[T any](err error) maybe[T] {
+	return maybe[T]{
+		err: err,
+	}
+}
+
+// Marshallable represents any TPM type that can be marshalled.
+type Marshallable interface {
+	// marshal will serialize the given value, appending onto the given buffer.
+	// Returns an error if the value is not marshallable.
+	marshal(buf *bytes.Buffer)
+}
+
+// Unmarshallable represents any TPM type that can be marshalled or unmarshalled.
+type Unmarshallable interface {
+	Marshallable
```

Comment:
Might be good to make it more clear the relationship between Marshallable and Unmarshallable